sycl/source/device_selector.cpp

```diff
@@ -157,5 +161,152 @@ int host_selector::operator()(const device &dev) const {
   return Score;
 }
 
+namespace ONEAPI {
+namespace detail {
+struct filter {
+  backend Backend;
+  RT::PiDeviceType DeviceType;
+  int DeviceNum;
+  bool hasBackend = false;
+  bool hasDeviceType = false;
+  bool hasDeviceNum = false;
+  int MatchesSeen = 0;
+};
+
+std::vector<std::string> tokenize(std::string filter, std::string delim) {
+  std::vector<std::string> Tokens;
+  size_t Pos = 0;
+  std::string Input = filter;
+  std::string Tok;
+
+  while ((Pos = Input.find(delim)) != std::string::npos) {
+    Tok = Input.substr(0, Pos);
+    Input.erase(0, Pos + delim.length());
+
+    if (!Tok.empty()) {
+      Tokens.push_back(Tok);
+    }
+  }
+
+  // Add remainder
+  if (!Input.empty())
+    Tokens.push_back(Input);
+
+  return Tokens;
+}
+
+filter create_filter(std::string Input) {
+  filter Result;
+  std::string Error = "Invalid filter string! Valid strings conform to "
+                      "BE:DeviceType:DeviceNum, where any are optional";
+
+  std::vector<std::string> Tokens = tokenize(Input, ":");
+  std::regex IntegerExpr("[[:digit:]]+");
+
+  // There should only be up to 3 tokens.
+  // BE:Device Type:Device Num
+  if (Tokens.size() > 3)
+    throw runtime_error(Error, PI_INVALID_VALUE);
+
+  for (const std::string &Token : Tokens) {
+    if (Token == "cpu" && !Result.hasDeviceType) {
+      Result.DeviceType = PI_DEVICE_TYPE_CPU;
+      Result.hasDeviceType = true;
+    } else if (Token == "gpu" && !Result.hasDeviceType) {
+      Result.DeviceType = PI_DEVICE_TYPE_GPU;
+      Result.hasDeviceType = true;
+    } else if (Token == "accelerator" && !Result.hasDeviceType) {
+      Result.DeviceType = PI_DEVICE_TYPE_ACC;
+      Result.hasDeviceType = true;
+    } else if (Token == "opencl" && !Result.hasBackend) {
+      Result.Backend = backend::opencl;
+      Result.hasBackend = true;
+    } else if (Token == "level-zero" && !Result.hasBackend) {
+      Result.Backend = backend::level_zero;
+      Result.hasBackend = true;
+    } else if (Token == "cuda" && !Result.hasBackend) {
+      Result.Backend = backend::cuda;
+      Result.hasBackend = true;
+    } else if (Token == "host") {
+      if (!Result.hasBackend) {
+        Result.Backend = backend::host;
+        Result.hasBackend = true;
+      } else if (!Result.hasDeviceType && Result.Backend != backend::host) {
+        // We already set everything earlier or it's an error.
+        throw runtime_error("Cannot specify host device with non-host backend.",
+                            PI_INVALID_VALUE);
+      }
+    } else if (std::regex_match(Token, IntegerExpr) && !Result.hasDeviceNum) {
+      Result.DeviceNum = std::stoi(Token);
```

Comment:
I think we need to catch exceptions that might be throw by `std::stoi` and convert them to sycl:: one.

Reply:
Ok